src/core/include/units/magnitude.h

```diff
@@ -0,0 +1,337 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Mateusz Pusz
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <units/ratio.h>
+#include <cstdint>
+#include <numbers>
+
+namespace units::mag
+{
+
+/**
+ * @brief  A basis vector in our magnitude representation, raised to some rational power.
+ *
+ * The set of basis vectors must be linearly independent: that is, no product of basis powers can ever equal 1, unless
+ * all exponents are zero.  To achieve this, we use arbitrarily many prime numbers as basis vectors, and also various
+ * irrational numbers such as pi.
+ *
+ * @tparam Base  A type representing the basis vector.  Must have a numeric static `value` member.
+ * @tparam Power  The rational power to which the base is raised.
+ */
+template<typename Base, ratio Power = ratio{1}>
+struct base_power {
+  using base = Base;
+  static constexpr ratio power = Power;
+};
+
+/**
+ * @brief  A type trait and concept to detect whether something is a valid "base power".
+ */
+template<typename T>
+struct is_base_power;
+template<typename T>
+inline constexpr bool is_base_power_v = is_base_power<T>::value;
+template<typename T>
+concept BasePower = is_base_power_v<T>;
+
+/**
+ * @brief  A representation for positive real numbers which optimizes taking products and rational powers.
```

Comment:
The new version checks the individual base powers for validity.  Since valid base powers always have a positive base, yes, we do now check this with a constraint.